File: lib/core/crypto/file_encryptor.dart
import 'dart:convert';
import 'dart:math';
import 'dart:typed_data';

import 'package:pointycastle/export.dart';

/// Encrypts and decrypts maFile data using AES-256-CBC with PBKDF2 key derivation.
/// Direct port of FileEncryptor.cs from C#.
///
/// Encryption scheme:
/// - PBKDF2 with 50,000 iterations (SHA1) to derive 32-byte key from password + salt
/// - AES-256 in CBC mode with PKCS7 padding
/// - 8-byte random salt, 16-byte random IV
class FileEncryptor {
  static const int _pbkdf2Iterations = 50000;
  static const int _saltLength = 8;
  static const int _keySizeBytes = 32;
  static const int _ivLength = 16;

  /// Returns an 8-byte cryptographically random salt as Base64 string.
  static String getRandomSalt() {
    final salt = _getRandomBytes(_saltLength);
    return base64.encode(salt);
  }

  /// Returns a 16-byte cryptographically random IV as Base64 string.
  static String getInitializationVector() {
    final iv = _getRandomBytes(_ivLength);
    return base64.encode(iv);
  }

  /// Derives a 32-byte encryption key from password and salt using PBKDF2-SHA1.
  static Uint8List _getEncryptionKey(String password, String salt) {
    if (password.isEmpty) throw ArgumentError('Password is empty');
    if (salt.isEmpty) throw ArgumentError('Salt is empty');

    final saltBytes = base64.decode(salt);
    final pbkdf2 = PBKDF2KeyDerivator(HMac(SHA1Digest(), 64))
      ..init(Pbkdf2Parameters(
        Uint8List.fromList(saltBytes),
        _pbkdf2Iterations,
        _keySizeBytes,
      ));
    return pbkdf2.process(Uint8List.fromList(utf8.encode(password)));
  }

  /// Encrypts plaintext using AES-256-CBC and returns Base64-encoded ciphertext.
  ///
  /// [password] - User's encryption password.
  /// [passwordSalt] - Base64-encoded 8-byte salt.
  /// [iv] - Base64-encoded 16-byte initialization vector.
  /// [plaintext] - The data to encrypt.
  static String encryptData(
      String password, String passwordSalt, String iv, String plaintext) {
    if (password.isEmpty) throw ArgumentError('Password is empty');
    if (passwordSalt.isEmpty) throw ArgumentError('Salt is empty');
    if (iv.isEmpty) throw ArgumentError('Initialization Vector is empty');
    if (plaintext.isEmpty) throw ArgumentError('Plaintext data is empty');

    final key = _getEncryptionKey(password, passwordSalt);
    final ivBytes = Uint8List.fromList(base64.decode(iv));

    final cipher = PaddedBlockCipher('AES/CBC/PKCS7')
      ..init(
        true,
        PaddedBlockCipherParameters(
          ParametersWithIV(KeyParameter(key), ivBytes),
          null,
        ),
      );

    final plaintextBytes = Uint8List.fromList(utf8.encode(plaintext));
    final ciphertext = cipher.process(plaintextBytes);
    return base64.encode(ciphertext);
  }

  /// Decrypts Base64-encoded ciphertext using AES-256-CBC.
  /// Returns plaintext string, or null if the password is wrong.
  ///
  /// [password] - User's encryption password.
  /// [passwordSalt] - Base64-encoded 8-byte salt.
  /// [iv] - Base64-encoded 16-byte initialization vector.
  /// [encryptedData] - Base64-encoded ciphertext.
  static String? decryptData(
      String password, String passwordSalt, String iv, String encryptedData) {
    if (password.isEmpty) throw ArgumentError('Password is empty');
    if (passwordSalt.isEmpty) throw ArgumentError('Salt is empty');
    if (iv.isEmpty) throw ArgumentError('Initialization Vector is empty');
    if (encryptedData.isEmpty) throw ArgumentError('Encrypted data is empty');

    final key = _getEncryptionKey(password, passwordSalt);
    final ivBytes = Uint8List.fromList(base64.decode(iv));
    final ciphertext = Uint8List.fromList(base64.decode(encryptedData));

    try {
      final cipher = PaddedBlockCipher('AES/CBC/PKCS7')
        ..init(
          false,
          PaddedBlockCipherParameters(
            ParametersWithIV(KeyParameter(key), ivBytes),
            null,
          ),
        );

      final plaintext = cipher.process(ciphertext);
      return utf8.decode(plaintext);
    } catch (_) {
      // Bad password yields a bad key → decryption error
      return null;
    }
  }

  static Uint8List _getRandomBytes(int length) {
    final random = Random.secure();
    return Uint8List.fromList(
        List.generate(length, (_) => random.nextInt(256)));
  }
}

File: telegram-bot/steamguard/file_encryptor.py
"""maFile encryption, compatible with the desktop app and the original C# SDA.

    PBKDF2 (SHA1, 50000 iterations, 8-byte salt) -> 32-byte key
    AES-256-CBC (PKCS7 padding, 16-byte IV)

Salt and IV live in manifest.json per entry; the ciphertext is base64 inside
the .maFile. A file encrypted here decrypts in the Flutter app and vice versa —
`tests/test_file_encryptor.py` pins that with a fixture produced by the Dart
implementation.
"""

from __future__ import annotations

import base64
import os

from cryptography.hazmat.primitives import hashes, padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC

PBKDF2_ITERATIONS = 50_000
SALT_LENGTH = 8
KEY_SIZE_BYTES = 32
IV_LENGTH = 16
_AES_BLOCK_BITS = 128


def get_random_salt() -> str:
    """A fresh 8-byte salt, base64-encoded."""
    return base64.b64encode(os.urandom(SALT_LENGTH)).decode("ascii")


def get_initialization_vector() -> str:
    """A fresh 16-byte IV, base64-encoded."""
    return base64.b64encode(os.urandom(IV_LENGTH)).decode("ascii")


def encrypt_data(password: str, salt: str, iv: str, plaintext: str) -> str:
    """Encrypt ``plaintext``, returning base64 ciphertext."""
    if not plaintext:
        raise ValueError("Plaintext data is empty")

    key = _derive_key(password, salt)
    iv_bytes = base64.b64decode(_require(iv, "Initialization Vector"))

    padder = padding.PKCS7(_AES_BLOCK_BITS).padder()
    padded = padder.update(plaintext.encode("utf-8")) + padder.finalize()

    encryptor = Cipher(algorithms.AES(key), modes.CBC(iv_bytes)).encryptor()
    ciphertext = encryptor.update(padded) + encryptor.finalize()

    return base64.b64encode(ciphertext).decode("ascii")


def decrypt_data(password: str, salt: str, iv: str, encrypted: str) -> str | None:
    """Decrypt base64 ciphertext, or return None if the password is wrong.

    A wrong password derives a wrong key, which surfaces as a padding or UTF-8
    failure rather than anything more specific — so both are treated as
    "bad passkey", matching the desktop app.
    """
    if not encrypted:
        raise ValueError("Encrypted data is empty")

    key = _derive_key(password, salt)
    iv_bytes = base64.b64decode(_require(iv, "Initialization Vector"))

    try:
        ciphertext = base64.b64decode(encrypted)
        decryptor = Cipher(algorithms.AES(key), modes.CBC(iv_bytes)).decryptor()
        padded = decryptor.update(ciphertext) + decryptor.finalize()

        unpadder = padding.PKCS7(_AES_BLOCK_BITS).unpadder()
        plaintext = unpadder.update(padded) + unpadder.finalize()
        return plaintext.decode("utf-8")
    except Exception:
        return None


def _derive_key(password: str, salt: str) -> bytes:
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA1(),
        length=KEY_SIZE_BYTES,
        salt=base64.b64decode(_require(salt, "Salt")),
        iterations=PBKDF2_ITERATIONS,
    )
    return kdf.derive(_require(password, "Password").encode("utf-8"))


def _require(value: str, name: str) -> str:
    if not value:
        raise ValueError(f"{name} is empty")
    return value
